File: backend/app/schemas/evidence.py
from __future__ import annotations

import datetime as dt

from pydantic import BaseModel, Field


class EvidenceEventCreate(BaseModel):
    """POST /api/v1/evidence 바디 — 일반 판단 기록 기록.

    action_id/approval_id/run_id는 받지 않는다: DB 트리거(trg_evidence_context_match)가
    "제공되면 같은 케이스 소속 행이 실제로 존재해야 함"을 강제하는데, 이 엔드포인트를 쓰는
    화면(RBAC·해석 확인·발송 실행 등)은 아직 그 참조 도메인(승인 결정·런)이 백엔드에
    배선되지 않아 실제 행을 보장할 수 없다(docs/DB_SCHEMA.md §4.5 R2.5 노트) — case_id만 받는다.
    actor(누가)는 세션에서 도출한다(요청 바디로 신뢰하지 않음, approvals.py와 동일 원칙).
    """

    type: str = Field(min_length=1)
    case_id: str | None = None
    summary: str = Field(min_length=1)
    input_hash: str | None = None
    output_hash: str | None = None
    trace_id: str | None = None
    request_id: str | None = None
    payload_ref: str | None = None


class EvidenceEventOut(BaseModel):
    id: str
    company_id: str
    event_no: int
    type: str
    at: dt.datetime
    case_id: str | None
    action_id: str | None
    approval_id: str | None
    run_id: str | None
    actor_type: str
    actor_user_id: str | None
    actor_display: str | None
    summary: str
    input_hash: str | None
    output_hash: str | None
    trace_id: str | None
    request_id: str | None
    payload_ref: str | None
    created_at: dt.datetime

    model_config = {"from_attributes": True}

File: src/lib/api/evidence.ts
import { useSessionStore } from '@/stores/sessionStore';
import type { EvidenceEvent, EvidenceType } from '@/types';
import { apiFetch } from './client';

// GET/POST /api/v1/evidence 어댑터(R2.5) — backend/app/schemas/evidence.py의 EvidenceEventOut
// 그대로(snake_case). docs/DB_SCHEMA.md §8 매핑 계약: hash=input_hash, evidenceRef='#'+event_no,
// actor=actor_display.
export interface EvidenceEventDto {
  id: string;
  company_id: string;
  event_no: number;
  type: string;
  at: string;
  case_id: string | null;
  summary: string;
  input_hash: string | null;
  actor_display: string | null;
}

export function toEvidenceEvent(dto: EvidenceEventDto): EvidenceEvent {
  return {
    id: dto.id,
    type: dto.type as EvidenceType,
    at: dto.at,
    caseId: dto.case_id ?? undefined,
    hash: dto.input_hash ?? undefined,
    summary: dto.summary,
    actor: dto.actor_display ?? undefined,
    evidenceRef: `#${dto.event_no}`,
  };
}

// evidenceStore.append()가 real 모드에서 호출하는 범용 기록 경로 — action_id/approval_id/run_id는
// 보내지 않는다(백엔드 서비스가 참조 도메인이 아직 실제 행을 보장 못 하는 경우가 많아 거부하므로,
// 여기서부터 그 필드들을 만들어 보내지 않는다. src/stores/evidenceStore.ts 주석 참조).
export async function createEvidenceEvent(event: Pick<EvidenceEvent, 'type' | 'caseId' | 'summary'>): Promise<void> {
  const token = useSessionStore.getState().token ?? undefined;
  await apiFetch<EvidenceEventDto>('/api/v1/evidence', {
    method: 'POST',
    token,
    body: { type: event.type, case_id: event.caseId ?? null, summary: event.summary ?? '' },
  });
}

export async function fetchEvidence(): Promise<EvidenceEvent[]> {
  const token = useSessionStore.getState().token ?? undefined;
  const dtos = await apiFetch<EvidenceEventDto[]>('/api/v1/evidence', { token });
  return dtos.map(toEvidenceEvent);
}
